internal/corazawaf/rule.go

```diff
@@ -285,16 +291,17 @@ func (r *Rule) doEvaluate(phase types.RulePhase, tx *Transaction, cache map[tran
 			matchedValues = append(matchedValues, matchedChainValues...)
 			nr = nr.Chain
 		}
-		// we need to add disruptive actions in the end, otherwise they would be triggered without their chains.
-		if tx.RuleEngine != types.RuleEngineDetectionOnly {
-			tx.DebugLogger().Debug().Int("rule_id", rid).Msg("Disrupting transaction by rule")
-			for _, a := range r.actions {
-				if a.Function.Type() == rules.ActionTypeDisruptive || a.Function.Type() == rules.ActionTypeFlow {
-					tx.DebugLogger().Debug().Int("rule_id", rid).Str("action", a.Name).Msg("Evaluating action for rule")
-					a.Function.Evaluate(r, tx)
-				}
-			}
 
+		for _, a := range r.actions {
+			if a.Function.Type() == rules.ActionTypeFlow {
+				// Flow actions are evaluated also if the rule engine is set to DetectionOnly
+				tx.DebugLogger().Debug().Int("rule_id", rid).Str("action", a.Name).Msg("Evaluating flow action for rule")
+				a.Function.Evaluate(r, tx)
+			} else if a.Function.Type() == rules.ActionTypeDisruptive && tx.RuleEngine == types.RuleEngineOn {
+				// The parser enforces that the disruptive action is just one per rule (if more than one, only the last one is kept)
```

Comment:
> if more than one, only the last one is kept

But we are evaluating them all, isn't? Is it worth?

Reply:
Also, shall we cover this condition with a unit test?

Reply:
The parser itself keeps inside the list of actions only one disruptive action (the last parsed). So only one disruptive action at this point will be evaluated